File: leetcode/0198_house_robber.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int rob(vector<int>& nums) {
        int current = 0, previous = 0;

        for (auto num : nums) {
            int tmp = current;
            current = max(current, previous + num);
            previous = tmp;
        }

        return current;
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {1, 2, 3, 1};

    assert(4 == solution.rob(nums));
}

File: leetcode/0198_house_robber.py
# -*- coding: utf-8 -*-


class Solution:
    def rob(self, nums):
        current, previous = 0, 0

        for num in nums:
            current, previous = max(previous + num, current), current

        return current


if __name__ == '__main__':
    solution = Solution()

    assert 4 == solution.rob([1, 2, 3, 1])
    assert 12 == solution.rob([2, 7, 9, 3, 1])
